File: ruby/OnetWebService.rb
require 'net/https'
require 'uri'
require 'cgi'
require 'json'

class OnetWebService

  def initialize(api_key)
    @headers = { 'Accept' => 'application/json',
                 'X-API-Key' => api_key,
                 'User-Agent' => 'ruby-OnetWebService/2.00 (bot)' }
    self.set_version
  end
  
  def set_version(version = nil)
    @url_root =
      if version.nil?
        'https://api-v2.onetcenter.org/'
      else
        "https://api-v#{version}.onetcenter.org/"
      end
  end
  
  def call(path, query = {})
    url = @url_root + path + '?'
    query.each do |key, value|
      Array(value).each do |v|
        url += CGI::escape(key.to_s) + '=' + CGI::escape(v.to_s) + '&'
      end
    end
    url.chop!
    
    uri = URI.parse(url)
    request = Net::HTTP::Get.new(uri.request_uri)
    @headers.each { |k, v| request[k] = v }

    response = nil
    begin
      Net::HTTP.start(uri.host, uri.port,
                      :use_ssl => true,
                      :open_timeout => 10,
                      :read_timeout => 10,
                      :ssl_timeout => 10) do |http|
        response = http.request(request)
      end
    rescue StandardError => e
      return { 'error' => "Call to #{url} failed with reason: #{e.inspect}" }
    end
    
    result = nil
    case response
    when Net::HTTPOK, Net::HTTPUnprocessableEntity
      result = JSON.parse(response.body)
    else
      if !response.code.nil?
        result = { 'error' => "Call to #{url} failed with error code #{response.code}" }
      else
        result = { 'error' => "Call to #{url} failed with unknown reason" }
      end
    end
    result
  end

end

File: python/OnetWebService.py
import urllib.request, urllib.parse, urllib.error
import urllib.request, urllib.error, urllib.parse
import base64
import json

class OnetWebService:
    
    def __init__(self, api_key):
        self._headers = {
            'User-Agent': 'python-OnetWebService/2.00 (bot)',
            'X-API-Key': api_key,
            'Accept': 'application/json' }
        self.set_version()
    
    def set_version(self, version = None):
        if version is None:
            self._url_root = 'https://api-v2.onetcenter.org/'
        else:
            self._url_root = 'https://api-v' + version + '.onetcenter.org/'
    
    def call(self, path, *query):
        try:
            url = self._url_root + path
            if len(query) > 0:
                url += '?' + urllib.parse.urlencode(query, True)
            req = urllib.request.Request(url, None, self._headers)
            handle = None
            try:
                handle = urllib.request.urlopen(req)
            except urllib.error.HTTPError as e:
                if e.code == 422:
                    try:
                        return json.load(e)
                    except json.JSONDecodeError:
                        return { 'error': 'Call to ' + url + ' failed to return valid JSON' }
                    except UnicodeDecodeError:
                        return { 'error': 'Call to ' + url + ' failed to return valid UTF-8' }
                else:
                    return { 'error': 'Call to ' + url + ' failed with error code ' + str(e.code) }
            except urllib.error.URLError as e:
                return { 'error': 'Call to ' + url + ' failed with reason: ' + str(e.reason) }
            code = handle.getcode()
            if (code != 200) and (code != 422):
                return { 'error': 'Call to ' + url + ' failed with error code ' + str(code),
                        'urllib2_info': handle }
            try:
                return json.load(handle)
            except json.JSONDecodeError:
                return { 'error': 'Call to ' + url + ' failed to return valid JSON' }
            except UnicodeDecodeError:
                return { 'error': 'Call to ' + url + ' failed to return valid UTF-8' }
        except Exception as e:
            return { 'error': 'Call failed with unexpected error', 'exception': e }
